### ISSUE TITLE
Question - System Debug reduced for Trigger logger?

### ISSUE THREAD

**Issue by mikbranchaud** [j-fischer/rflib#23]
We are using both the trigger framework and Logger.  In my org level settings, I have System Debug Log Level set to INFO.  When running a test class, I have noticed that the data setup which fires a lot of triggers results in USER_DEBUG events in the log from logic in the various rflib trigger/feature switch classes even though it appears the respective log level of these statements is DEBUG.

Is this expected?  Is there a way to suppress these DEBUG statements as my logs for a test execution are now mostly filled with Debug statements originating from the trigger and feature framework.

**Comment by kurtfairfield** [comment]
have a look at this issue -> https://github.com/j-fischer/rflib/issues/6
perhaps it addresses you issue?

**Comment by j-fischer** [comment]
Hi @mikbranchaud ,

I apologize for the late reply, I've been heads down at work and missed any notification emails from Github. 

Unit tests do not see your current org configuration when it comes to custom settings, hence the default values will be applied. In the case of the System Debug Log Level, the default value is DEBUG, which is the reason why you are seeing all those statements in your log file. 

What I do in my projects to solve this to add a Utility class for testing purposes with a method that will insert global defaults for the Logger Settings. I call this method first thing in the `@TestSetup` method or first thing in the unit test method if no setup method is present.

Here is the example class that is used right in the project. Just copy that and adjust the settings to whatever you need. My recommendation is to generally turn of all reporting log levels by setting them to NONE.

https://github.com/j-fischer/rflib/blob/master/rflib/test/default/classes/rflib_TestUtil.cls

Hope that helps! Again, sorry for the delay. 

Johannes

cc: @kurtfairfield

**Comment by mikbranchaud** [comment]
Thanks for getting back to me.  The utility explains a lot.  I think it leaves a gap though unless I am missing something.  In a test centered development model, the first time an apex class runs is via the test class.  So having a lot of LOGGER details in a service class at the 'correct' (production appropriate) level of logging is initially designed so that it also helps troubleshoot the development.  Therefore I am frequently using DEBUG/TRACE level inside loops and reserving INFO for general information.  Since all the Framework logging is also done at DEBUG, this means that reducing the system logging to INFO or NONE means I would not see the logging I need to troubleshoot the development.  

The alternative is to not use the LOGGER utility to troubleshoot development (I could only use System.debug statements during development/testing then delete them all when done) or set it up in the service class with a higher level of detail like INFO/WARN and then change them all back after development is complete.  Since we are always adding features, this is a cyclical issue and would be a missed opportunity.

It seems a great enhancement would to be to just raise the threshold for utility based logging to DETAIL (I am not sure if your first post in this thread indicates this is not doable or not).  Bigger picture it might even be nice to introduce a new level of detail below that of DETAIL that is reserved for Only the utility to troubleshoot itself like UTILITY.

**Comment by j-fischer** [comment]
Hi @mikbranchaud ,

I think I am getting some of your issues, but I am not sure if I fully understand it, so let me add a some more comments and see if this helps: 

First, my general guidance is to use INFO as the most common logging level, the reason for this is some constraints on the reporting configuration side as well as the fact that I believe that it is the right level that gives you enough balance to introduce two more levels of details below (DEBUG & TRACE) that allows to add more context in a selective way. At this time, the log levels are in line with other log frameworks and I do not feel that there is any benefit of introducing another log level. In the case of your service class, I definitely would consider to have most of your log statements at the INFO level. That's what I do in my projects.

I don't like the idea of you having to add System.debug statements to your code and later remove it. It's a workaround, but that is not what the framework was meant for. The default System Debug Log Level is set to DEBUG because the idea is to mimic statements of System.debug which I commonly see in the field. By replacing your System.debug statements with RFLIB, you should have the same behaviour out of the box when it comes to the Apex Logs compared to what you had before. 

Now, having said all of that, you seem to have some challenges that are specific to unit tests. Based on what I understand, you do not like debug log statements to be printed when your set up a test case, but when the test runs. If switching the standard log level to INFO does not already help, there is some Apex behaviour that could do the trick. If you use the `@TestSetup` annotation for your test setup, you could use the utility method described in my previous post to prevent any log statements from being added to the Apex logs. But this behaviour would obviously carry over into your test case, which is not desired. But Apex actually reinitializes the runtime context between test setup and test method, which means that all static variables should be reset and re-initialized. That means that you could have a second utility method that updates the Custom Settings back to the defaults or whatever levels best support your development/debugging process. 
So, in other words, have one utility method to call in the first line of your `@TestSetup` method, and another one to call in the first line of your `@isTest` method. That should give you the results you are looking for. 

I haven't tested this out, so I might be missing something here and it won't work, but maybe give it a try and let me know what how it goes. 

Hope this helps! 

Cheers,
Johannes

**Comment by mikbranchaud** [comment]
I agree with your comments regarding ideal logging levels when using the framework in NEW custom code.  

To be precise, the enhancement that I think would be useful (both inside and outside of unit testing) would be to control the log level of the Trigger Framework itself.  While it is useful to see the DEBUG level logs when the framework finds a trigger and executes it, I find that there are a lot of log entries outputted by the framework that are boilerplate.  For every object on which DML is run with a trigger here is what I see:

1. 11:31:16:784 USER_DEBUG [375]|DEBUG|2021-04-08T01:31:17|DEBUG|0053R000000IKUrQAO|rflib_TriggerManager|0 active trigger(s) found for operation BEFORE_INSERT on object type Alias__c: 
2. 11:31:16:788 USER_DEBUG [375]|DEBUG|2021-04-08T01:31:17|DEBUG|0053R000000IKUrQAO|rflib_FeatureSwitch|Evaluating feature switch All_Triggers
3. 11:31:16:798 USER_DEBUG [375]|DEBUG|2021-04-08T01:31:17|DEBUG|0053R000000IKUrQAO|rflib_FeatureSwitch|All feature switch values: [{"attributes":{"type":"rflib_Feature_Switch__mdt","url":"/services/data/v51.0/sobjects/rflib_Feature_Switch__mdt/m013R0000000051QAA"},"Switch_Name__c":"All_Triggers","Scope_Type__c":"Global","Turned_On__c":true,"Id":"m013R0000000051QAA"}]
4. 11:31:16:809 USER_DEBUG [375]|DEBUG|2021-04-08T01:31:17|DEBUG|0053R000000IKUrQAO|rflib_FeatureSwitch|Feature switch All_Triggers is turned off? false
5. 11:31:16:809 USER_DEBUG [408]|INFO|There is no log event to be published; exiting.

or sometimes:
5. 11:31:16:890 USER_DEBUG [375]|DEBUG|2021-04-08T01:31:17|DEBUG|0053R000000IKUrQAO|rflib_TriggerManager|Is consecutive run? false

The first is nice since we are using the trigger framework.  Then the feature switch entries are not relevant to my project since we only have the one seed feature switch for the trigger framework itself.  Then there is an INFO entry from the logging framework: "There is no log event to be published; exiting" which I am not really clear about the source. Then there is a DEBUG entry about consecutive runs. We currently do not have any logic implemented for consecutive runs but I see the use of this (ideally would only display if true but whatever).  I would prefer these last entries were at TRACE level (or all of them if it was hard to disaggregate, or a different level per framework).   Maybe these would be useful in some contexts so from the framework perspective it would be ideal if this was configurable.

In any case, I like the approach you have suggested of using a log settings setter utility in setup/test class and another to reset.  This should really help.  I just want to point out for other readers (it took me a while to figure this out) that if you copy the rflib_TestUtil class, to make it actually log details during a test execution you need to change the Factory/Timer classes from the mock* to the Default*.  Interestingly, even though I set all the same values that TestUtil sets to 'NONE', I still see plenty of INFO entries for:
"There is no log event to be published; exiting"
Not a deal breaker but seems like this should not happen.

**Comment by j-fischer** [comment]
Hi @mikbranchaud, sorry that it took me a few days, but I needed some time to think things through and review options. 

Based on what you wrote, I agree that certain log levels do not seem appropriate and we could do without them. I have therefore made some changes that I would like you to review and let me know what you think about it. Would you mind taking a look at https://github.com/j-fischer/rflib/pull/25 and add your comments? 

Here are the reasons why I opted against making the log level configurable:
- There were only a handful messages, adding a config option just for those seemed a bit heavy
- The log message within the framework can only be TRACE or DEBUG so that they don't conflict with the reporting settings, which could cause infinite loops, so there would not really be much of an option to choose from. 
- The log messages that I left in DEBUG I find very valuable to be present using the default settings out of the box.

I hope those reasons make sense. 

Last, the following message `There is no log event to be published; exiting.` was created by the `BatchPlatformEventExecutor` and is a `System.debug()` message by design. This means that you are most likely an early adopter of the framework, thanks for that. What I can recommend for this is to switch the logger to a non-batch version. With the Winter 21 release, Salesforce gave Platform Events a dedicated Governor limit so that they conflict with DML Governor limits. This made the batched logger almost obsolete, I would only recommend its use for scenarios where one tries to reduce the publishing of competing Platform Events. 
Unless that is the case in your code base, you could consider changing the logger type to reduce this statement from your logs. 

Looking forward to getting your feedback on the PR. It's a small change but I hope it will help. 

Cheers,
Johannes

**Comment by j-fischer** [comment]
Hi @mikbranchaud ,

just to close the loop, new package versions have been published and contain the changes from the PR you looked at:

- RFLIB v2.4.0
- RFLIB-FS v1.2.0
- RFLIB-TF v1.2.0

Thanks,
Johannes
